targets/XPlatCppSdk/source/code/source/playfab/OneDSWinHttpPlugin.cpp

```diff
@@ -0,0 +1,111 @@
+
+namespace PlayFab
+{
+    std::string OneDSWinHttpPlugin::GetUrl(CallRequestContainer& requestContainer) const
+    {
+        return std::string("https://self.events.data.microsoft.com/OneCollector/1.0/");
+    }
+
+    void OneDSWinHttpPlugin::SetPredefinedHeaders(CallRequestContainer& requestContainer, HINTERNET hRequest)
+    {
+        OneDSCallRequestContainer& reqContainer = (OneDSCallRequestContainer&)requestContainer;
+
+        WinHttpAddRequestHeaders(hRequest, L"Accept: */*", -1, 0);
+        WinHttpAddRequestHeaders(hRequest, L"Client-Id: NO_AUTH", -1, 0);
+        WinHttpAddRequestHeaders(hRequest, L"Content-Type: application/bond-compact-binary", -1, 0);
+        WinHttpAddRequestHeaders(hRequest, L"Expect: 100-continue", -1, 0);
+        WinHttpAddRequestHeaders(hRequest, L"SDK-Version: EVT-PlayFab-XPlat-C++-No-3.0.275.1", -1, 0);
+        int64_t now = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+        WinHttpAddRequestHeaders(hRequest, (L"Upload-Time: " + std::to_wstring(now)).c_str(), -1, 0);
+        WinHttpAddRequestHeaders(hRequest, (L"Content-Length: " + std::to_wstring(reqContainer.requestBinaryBody.size())).c_str(), -1, 0);
+        WinHttpAddRequestHeaders(hRequest, L"Connection: Keep-Alive", -1, 0);
+        WinHttpAddRequestHeaders(hRequest, L"Cache-Control: no-cache", -1, 0);
+    }
+
+    bool OneDSWinHttpPlugin::GetBinaryPayload(CallRequestContainer& requestContainer, LPVOID& payload, DWORD& payloadSize) const
+    {
+        OneDSCallRequestContainer& reqContainer = (OneDSCallRequestContainer&)requestContainer;
+        payloadSize = (DWORD)reqContainer.requestBinaryBody.size();
+        payload = reqContainer.requestBinaryBody.data();
+        return true;
+    }
+
+    void OneDSWinHttpPlugin::ProcessResponse(CallRequestContainer& reqContainer, const int httpCode)
+    {
+        if ((httpCode >= 200 && httpCode < 300) || httpCode == 0)
+        {
+            // following One-DS recommendations, HTTP response codes within [200, 300) are success
+            // (0 means HTTP code could not be determined and we attempt to parse response)
+            Json::CharReaderBuilder jsonReaderFactory;
+            std::unique_ptr<Json::CharReader> jsonReader(jsonReaderFactory.newCharReader());
+            JSONCPP_STRING jsonParseErrors;
+            const bool parsedSuccessfully = jsonReader->parse(reqContainer.responseString.c_str(), reqContainer.responseString.c_str() + reqContainer.responseString.length(), &reqContainer.responseJson, &jsonParseErrors);
+
+            if (parsedSuccessfully)
+            {
+                auto result = reqContainer.responseJson.get("acc", Json::Value::null).asInt();
+                if (result > 0)
+                {
+                    // fully successful response
+                    reqContainer.errorWrapper.HttpCode = httpCode != 0 ? httpCode : 200;
```

Comment:
[Curious] : In what scenario can the returned HttpCode be 0?

Reply:
I wanted to be super-safe here and return a real HTTP code when possible, but also fallback to what application considers a successful call in situations when a real HTTP code was not returned (for example because WinHttpQueryHeaders API call failed in some implicit way). The HTTP call itself is not necessarily a failure, in this situation, it just we didn't get its response code, since it's a separate API. If we get to the point that we got "a response" in the form of "some result content" from the server (and not just an HTTP error code), then chances are close to 99% that it was 200.

---
In reply to: [244869709](https://github.com/PlayFab/SDKGenerator/pull/368#discussion_r244869709) [](ancestors = 244869709)